pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java

```diff
@@ -0,0 +1,358 @@
+    CONNECTED_TO_BOOTSTRAP_NODE,
+    FETCHING_LEADER_INFORMATION,
+    CONNECTING_TO_PARTITION_LEADER,
+    CONNECTED_TO_PARTITION_LEADER
+  }
+
+  private State _currentState;
+
+  private final boolean _metadataOnlyConsumer;
+  private final String _topic;
+  private final int _partition;
+  private final String _bootstrapNodes;
+  private SimpleConsumer _simpleConsumer;
+
+  private SimpleConsumerWrapper(String bootstrapNodes) {
+    _bootstrapNodes = bootstrapNodes;
+    _metadataOnlyConsumer = true;
+    _simpleConsumer = null;
+
+    // Topic and partition are ignored for metadata-only consumers
+    _topic = null;
+    _partition = Integer.MIN_VALUE;
+
+    setCurrentState(new ConnectingToBootstrapNode());
+  }
+
+  private SimpleConsumerWrapper(String bootstrapNodes, String topic, int partition) {
+    _bootstrapNodes = bootstrapNodes;
+    _topic = topic;
+    _partition = partition;
+    _metadataOnlyConsumer = false;
+    _simpleConsumer = null;
+
+    setCurrentState(new ConnectingToBootstrapNode());
+  }
+
+  private abstract class State {
+    private ConsumerState stateValue;
+
+    protected State(ConsumerState stateValue) {
+      this.stateValue = stateValue;
+    }
+
+    abstract void think();
+
+    abstract boolean isConnectedToKafkaBroker();
+
+    void handleConsumerException(Exception e) {
+      // TODO Remove this
+      System.out.println("e = " + e);
+      e.printStackTrace();
+
+      // By default, just log the exception and switch back to CONNECTING_TO_BOOTSTRAP_NODE (which will take care of
+      // closing the connection if it exists)
+      LOGGER.warn("Caught Kafka consumer exception, disconnecting and trying again", e);
+
+      setCurrentState(new ConnectingToBootstrapNode());
+    }
+
+    ConsumerState getStateValue() {
+      return stateValue;
+    }
+  }
+
+  private class ConnectingToBootstrapNode extends State {
+    public ConnectingToBootstrapNode() {
+      super(ConsumerState.CONNECTING_TO_BOOTSTRAP_NODE);
+    }
+
+    @Override
+    public void think() {
+      // Connect to a random bootstrap node
+      if (_simpleConsumer != null) {
+        try {
+          _simpleConsumer.close();
+        } catch (Exception e) {
+          LOGGER.warn("Caught exception while closing consumer, ignoring", e);
+        }
+      }
+
+      // TODO Parse bootstrapNodes
+      // TODO Remove magic numbers
+      try {
+        _simpleConsumer = new SimpleConsumer("localhost", KafkaStarterUtils.DEFAULT_KAFKA_PORT, 10000, 10000, "potato");
+        setCurrentState(new ConnectedToBootstrapNode());
+      } catch (Exception e) {
+        handleConsumerException(e);
+      }
+    }
+
+    @Override
+    boolean isConnectedToKafkaBroker() {
+      return false;
+    }
+  }
+
+  private class ConnectedToBootstrapNode extends State {
+    protected ConnectedToBootstrapNode() {
+      super(ConsumerState.CONNECTED_TO_BOOTSTRAP_NODE);
+    }
+
+    @Override
+    void think() {
+      if (_metadataOnlyConsumer) {
```

Comment:
i am tempted to split the metadata only consumer as a separate class with 2 states